include/dav1d/headers.rs

```diff
@@ -2096,6 +2096,48 @@ pub struct Dav1dFrameHeader_segmentation {
     pub qidx: [u8; DAV1D_MAX_SEGMENTS as usize],
 }
 
+/// Like a `[bool; RAV1D_MAX_SEGMENTS as usize]`,
+/// but the [`bool`]s are compressed into the bits of the [`u8`].
+#[derive(Clone, Copy, Default)]
+pub struct LossLess {
+    bits: u8,
+}
+
+impl LossLess {
+    pub const fn empty() -> Self {
+        Self { bits: 0 }
+    }
+
+    pub const fn get(&self, index: u8) -> bool {
+        const { assert!(u8::BITS >= RAV1D_MAX_SEGMENTS as _) };
+        self.bits & (1 << index) != 0
+    }
+
+    pub const fn all(&self) -> bool {
+        self.bits == ((1u64 << RAV1D_MAX_SEGMENTS) - 1) as u8
+    }
+
+    pub const fn to_array(&self) -> [bool; RAV1D_MAX_SEGMENTS as usize] {
+        let mut a = [false; RAV1D_MAX_SEGMENTS as usize];
+        let mut i = 0;
+        while i < RAV1D_MAX_SEGMENTS {
+            a[i as usize] = self.get(i);
+            i += 1;
+        }
+        a
+    }
+
+    pub fn from_array(a: [bool; RAV1D_MAX_SEGMENTS as usize]) -> Self {
+        let mut this = Self::empty();
+        let mut i = 0;
+        while i < RAV1D_MAX_SEGMENTS {
+            this.bits |= (a[i as usize] as u8) << i;
+            i += 1;
+        }
+        this
+    }
```

Comment:
These two methods can be replaced with instructions `pdep` and `pext` (slow on `Zen` processor).
And there are other [alternatives](https://stackoverflow.com/questions/8461126/how-to-create-a-byte-out-of-8-bool-values-and-vice-versa#answer-51750902), just for reference, since they are unsafe operations.

Reply:
Interesting... Thanks for letting me know!  These `fn`s are only called a few times in `DRav1d` conversions, and in `mod obu`, though, none of which are hot, so I'm not sure if it's worth it to make this optimization.  Definitely not asm ones.  For the safe multiplication one, maybe, but I'm not sure it's worth the clarity if the `fn` is not hot at all.  Maybe I'll do it.

Reply:
I would say stick with the simpler/clearer loop-based approach for now. If we decide to go with this change (which we may not), we can profile and use a more optimized approach if it proves to be on the hot path.

Reply:
Left a comment in 2720efd explaining this for if we do want to optimize it in the future.